ddspipe_core/src/cpp/monitoring/producers/TopicsMonitorProducer.cpp

```diff
@@ -0,0 +1,333 @@
+}
+
+void TopicsMonitorProducer::type_discovered(
+        const std::string& topic_name,
+        const std::string& type_name)
+{
+    if (!enabled_)
+    {
+        // Don't save the data if the producer is not enabled
+        return;
+    }
+
+    // Take the lock to prevent:
+    //      1. Changing the data while it's being saved.
+    //      2. Simultaneous calls to msg_lost.
+    std::lock_guard<std::mutex> lock(mutex_);
+
+    if (!topic_data_.count(topic_name))
+    {
+        topic_data_[topic_name].name(topic_name);
+        topic_data_[topic_name].type_name(type_name);
+    }
+
+    topic_data_[topic_name].type_discovered(true);
+}
+
+void TopicsMonitorProducer::type_mismatch(
+        const types::DdsTopic& topic)
+{
+    if (!enabled_)
+    {
+        // Don't save the data if the producer is not enabled
+        return;
+    }
+
+    // Take the lock to prevent:
+    //      1. Changing the data while it's being saved.
+    //      2. Simultaneous calls to msg_lost.
+    std::lock_guard<std::mutex> lock(mutex_);
+
+    const auto& topic_name = topic.m_topic_name;
+    const auto& type_name = topic.type_name;
+
+    if (!topic_data_.count(topic_name))
+    {
+        topic_data_[topic_name].name(topic_name);
+        topic_data_[topic_name].type_name(type_name);
+    }
+
+    topic_data_[topic.m_topic_name].type_mismatch(true);
+}
+
+void TopicsMonitorProducer::qos_mismatch(
+        const types::DdsTopic& topic)
+{
+    if (!enabled_)
+    {
+        // Don't save the data if the producer is not enabled
+        return;
+    }
+
+    // Take the lock to prevent:
+    //      1. Changing the data while it's being saved.
+    //      2. Simultaneous calls to msg_lost.
+    std::lock_guard<std::mutex> lock(mutex_);
+
+    const auto& topic_name = topic.m_topic_name;
+    const auto& type_name = topic.type_name;
+
+    if (!topic_data_.count(topic_name))
+    {
+        topic_data_[topic_name].name(topic_name);
+        topic_data_[topic_name].type_name(type_name);
+    }
+
+    topic_data_[topic_name].qos_mismatch(true);
+}
+
+MonitoringTopics TopicsMonitorProducer::save_data_()
+{
+    // Take the lock to prevent saving the data while it's changing
+    std::lock_guard<std::mutex> lock(mutex_);
+
+    std::vector<DdsTopic> topics_data;
+
+    // Iterate through the different topics
+    for (auto& topic : participant_data_)
+    {
+        auto& dds_topic = topic.first;
+        auto& participants = topic.second;
+
+        std::vector<DdsTopicData> topic_participants;
+
+        for (auto& participant : participants)
+        {
+            auto& participant_id = participant.first;
+            auto& participant_data = participant.second;
+
+            // Calculate the message reception frequency
+            const double period_in_secs = (double) period / 1000;
+            participant_data.frequency((double) participant_data.msgs_received() / period_in_secs);
```

Comment:
NIT: I'm not sure frequency is the right term (I tend to expect hertzs for its unit). How about reception rate for example?

Reply:
I could settle for `msg_reception_rate`. Deal?